lib/model.js

```diff
@@ -2530,163 +2530,313 @@ class Model {
+              }
+            }
+            if (updatedAtAttr && !values[updatedAtAttr]) {
+              values[updatedAtAttr] = now;
+              if (!options.fields.includes(updatedAtAttr)) {
+                options.fields.push(updatedAtAttr);
+              }
             }
+
+            const out = Object.assign({}, Utils.mapValueFieldNames(values, options.fields, model));
+            for (const key of model._virtualAttributes) {
+              delete out[key];
+            }
+            return out;
           });
-        }
-        return results;
-      });
-    }).then(() => {
-      // map fields back to attributes
-      instances.forEach(instance => {
-        for (const attr in this.rawAttributes) {
-          if (this.rawAttributes[attr].field &&
-              instance.dataValues[this.rawAttributes[attr].field] !== undefined &&
-              this.rawAttributes[attr].field !== attr
-          ) {
-            instance.dataValues[attr] = instance.dataValues[this.rawAttributes[attr].field];
-            delete instance.dataValues[this.rawAttributes[attr].field];
+
+          // Map attributes to fields for serial identification
+          const fieldMappedAttributes = {};
+          for (const attr in model.tableAttributes) {
+            fieldMappedAttributes[model.rawAttributes[attr].field || attr] = model.rawAttributes[attr];
+          }
+
+          // Map updateOnDuplicate attributes to fields
+          if (options.updateOnDuplicate) {
+            options.updateOnDuplicate = options.updateOnDuplicate.map(attr => model.rawAttributes[attr].field || attr);
+            // Get primary keys for postgres to enable updateOnDuplicate
+            options.upsertKeys = _.chain(model.primaryKeys).values().map('fieldName').value();
+            if (Object.keys(model.uniqueKeys).length > 0) {
+              options.upsertKeys = _.chain(model.uniqueKeys).values().filter(c => c.fields.length === 1).map('column').value();
+            }
+          }
+
+          // Map returning attributes to fields
+          if (options.returning && Array.isArray(options.returning)) {
+            options.returning = options.returning.map(attr => model.rawAttributes[attr].field || attr);
           }
-          instance._previousDataValues[attr] = instance.dataValues[attr];
-          instance.changed(attr, false);
+
+          return model.QueryInterface.bulkInsert(model.getTableName(options), records, options, fieldMappedAttributes).then(results => {
+            if (Array.isArray(results)) {
+              results.forEach((result, i) => {
+                const instance = instances[i];
+
+                for (const key in result) {
+                  if (!instance || key === model.primaryKeyAttribute &&
+                    instance.get(model.primaryKeyAttribute) &&
+                    ['mysql', 'mariadb', 'sqlite'].includes(dialect)) {
+                    // The query.js for these DBs is blind, it autoincrements the
```

Comment:
For MySQL I have seen this behaviour with updateOnDuplicate https://github.com/sequelize/sequelize/issues/11223#issuecomment-515669115

Reply:
> I don't think we can reliably support this unless MySQL adds support for RETURNING * equivalent

That is exactly the root of the problem. PostgreSQL and SQL Server both support `RETURNING` and therefore don't have any issue.